Various Cycles 4.2 improvements

murraystevenson opened the pull request:
This brings a few changes related to the recent Cycles update:

- Exposing the `is_sphere` socket on point and spot lights to allow reverting them back to the non-physically based behaviour from Cycles 3.6.0 and earlier. This is equivalent to Blender's [Soft Falloff](https://developer.blender.org/docs/release_notes/4.1/rendering/#lights) option, though the logic in Blender is reversed for presentation.
- Changing over to Cycles' new [blue noise](https://developer.blender.org/docs/release_notes/4.2/cycles/#sampling) based sampling pattern. Blender provides the ability to revert back to the "Classic" sampling pattern as an advanced option. We could add this option back to CyclesOptions if necessary later, but it seems worthwhile to at least introduce this change of the default in advance of Gaffer 1.5.
- Removing our custom handing of unnormalized lights now that this is handled by Cycles. This fixes an issue where scaling wasn't being taken into account for unnormalized quad and disk lights with our code, but does introduce a breaking look change as the strength of unnormalized point, spot and disk lights differ with the Cycles code. Unnormalized disk lights get closer to Arnold as a result, but point and spot lights diverge as the Cycles code is based on a sphere with a surface area of 1, while Arnold and our code is/was based on a surface area of 4.

These changes are grouped in a single PR in order to more easily facilitate user testing with a single build artifact.

johnhaddon commented:
> point and spot lights diverge as the Cycles code is based on a sphere with a surface area of 1, while Arnold and our code is/was based on a surface area of 4.

I can see the benefit in matching Blender here, but it sounds like we'll also have lost the match between UsdLux lights in Arnold and Cycles. Perhaps it's worth keeping that?

Otherwise, LGTM.

murraystevenson commented:
Rebased to fix the Changes.md conflict and added a [commit](https://github.com/GafferHQ/gaffer/pull/6061/commits/a4a1caa56256c99b9b1e9a2ae3df0673fb56a551) that scales intensity of Quad, Disk, Spot and Point lights by `pi` to match behaviour introduced in [Blender](https://projects.blender.org/blender/blender/src/commit/05c97df2dead07630d7adf7dabe8810afcc903c3/source/blender/io/usd/intern/usd_reader_light.cc#L89-L92)/[hdCycles](
https://projects.blender.org/blender/blender/src/commit/05c97df2dead07630d7adf7dabe8810afcc903c3/intern/cycles/hydra/light.cpp#L101-L104). This change also further matches intensity with the equivalent Arnold lights. As this would be a breaking change in look for existing projects updating to Gaffer 1.5, I've added a way to disable this by setting the `GAFFERCYCLES_USE_LEGACY_LIGHTS` environment variable to `1`.

We could go further to attempt to standardise light behaviour, but it gets a bit murky without a more defined UsdLuxLight specification. The remaining differences are documented in the commit for the benefit of the next brave soul to venture into this oddly dark corner...
